lib/task_manager.js

```diff
@@ -0,0 +1,66 @@
+class TaskManager {
+  constructor() {
+    this.taskData = {};
+    this.queue;
+    this.buffer = [];
+    this.lastPushedToBuffer = 0;
+  }
+
+  async initQueue(maxConcurrentRequests) {
+    const PQueue = (await import('p-queue')).default;
+    this.queue = new PQueue({ concurrency: maxConcurrentRequests });
+    this.queue.on('completed', (data) => this.handleNewData(data));
+  }
+
+  clearQueue() {
+    this.queue.pause();
+    this.queue.clear();
+  }
+
+  #generateDefensiveCopy() {
+    const bufferCopy = [];
+    for (const data of this.buffer) bufferCopy.push(data);
+    return bufferCopy;
+  }
+
+  #shiftOriginalBuffer(shiftsCount) {
+    while (shiftsCount >= 0) {
+      this.buffer.shift();
+      shiftsCount--;
+    }
+  }
+
+  retrieveCompleted() {
+    const bufferCopy =  this.#generateDefensiveCopy();
+    let shiftsCount = bufferCopy.length - 1;
+    this.#shiftOriginalBuffer(shiftsCount);
+
+    return bufferCopy;
+  }
+
+  #pushAllEligable() {
+    while (this.taskData[this.lastPushedToBuffer]) {
+      for (const data of this.taskData[this.lastPushedToBuffer]) {
+        this.buffer.push(data);
+        this.lastProcessedPosition = {
```

Comment:
This is getting messy, we have 'lastProcessed' and 'lastBuffered'. However here you are mixing them here, the fact that something is eligible for going in the buffer doesn't make it any more 'processed', it still needs to be consumed from the Kafka side and send through the producer.

Reply:
You don't need this member variable. The user of this class can see what the last value is once it calls `retrieveCompleted`